provider/bidengine/order.go

```diff
@@ -150,44 +152,52 @@ loop:
 			res := result.Value().(dtypes.Group)
 			group = &res
 
-			if !o.shouldBid(group) {
+			if o.bid == nil && !o.shouldBid(group) {
 				break
```

Comment:
Good catch.   This part is tricky.  If `o.bid` is populated that means we restarted the process and recovered the bid from a previous execution.  We should validate that it can still be reserved, and close the bid if it can't be.

Also, if we can no longer fulfill the group (`o.shouldBid()` is false), we should cancel the original bid anyways.

So I think that the right thing to do is:

1. continue reserving space if `shouldBid()` is true.
1. exit the main loop if `shouldBid()` is false.
1. when cleaning up, cancel the bid if we exited before knowing if we won the lease.
1. when cleaning up, unconditionally unreserve space if we did not win the lease.  I think that `Unreserve` method signature can be changed to only require the order ID. 

Thoughts?

Reply:
To close the bid I'll need to send a `mtypes.MsgCloseBid` I am presuming?

Reply:
I'm sending a `*mtypes.MsgCloseBid` now as well as the other changes you recommended.